feat(api): (2) Upgrade upload APIs to Version 2

Akashsah2003 opened the pull request:
<!-- SPDX-FileCopyrightText: © Fossology contributors

     SPDX-License-Identifier: GPL-2.0-only
-->

<!-- Please refer to CONTRIBUTING.md (https://github.com/fossology/fossology/blob/master/CONTRIBUTING.md)
before creating the pull request to make sure you follow all the standards. -->

## Description

This PR aims to do tasks 7, 8, 12, 14, 15, 16, 17 of issue #2623.

### Changes

1. **Moved `groupName` variable from headers to query parameters**

2. **/updoads**, POST
- Moved `folderId` , `uploadDescription`, `public`, `applyGlobal`, `ignoreScm`, `uploadType` variables from the headers to body parameters.
- Changed unit test for the controller.
3. **/uploads/{id}/agents**, GET
- Changed these fields to camelCase: `agent_id`, `agent_rev`, `agent_name`
4. **/uploads/{id}/licenses/edited**, GET
- Changed these fields to camelCase: `spdx_id`
5. **/uploads/{id}/licenses/scanned**, GET
- Changed these fields to camelCase: `shortname`
6. **/uploads/{id}/item/{itemId}/tree/view**, GET
- Moved `page` , `limit` variables from the headers to query parameters.



## How to test

- Changes pass all the unit tests.
- Can be tested manually using postman or swagger.

/cc @dushimsam @GMishx @shaheemazmalmmd

dushimsam commented:
Hello @Akashsah2003 , these changes should go a long with updating the API documentation of the version-2 APIs

Akashsah2003 commented:
Hello @dushimsam, changes made to documentation. Please take a look.

GMishx commented on src/www/ui/api/Controllers/UploadController.php:
```diff
@@ -417,11 +417,11 @@ private function changeUpload($request, $response, $args, $isCopy)
   public function postUpload($request, $response, $args)
   {
-    $uploadType = $request->getHeaderLine('uploadType');
+    $reqBody = $this->getParsedBody($request);
+    $uploadType = $reqBody['uploadType'] ?? null;
```
The API should still support V1 requests in deprecated mode until it is completely removed. You can read the version of request with `ApiVersion::getVersion($request) == ApiVersion::V2`.

See https://github.com/fossology/fossology/pull/2617/files#diff-1bae1ab555e908fff79a5b02ff0f05677a8d1f5653259e98531071acfeb7e16bR42 for reference.

GMishx commented on src/www/ui/api/Controllers/UploadController.php:
```diff
@@ -1093,7 +1093,13 @@ public function getAllAgents($request, $response, $args)
+        $agent['agentId'] = intval($agent['agent_id']);
+        $agent['agentRev'] = $agent['agent_rev'];
+        $agent['agentName'] = $agent['agent_name'];
+        unset($agent['agent_id']);
+        unset($agent['agent_name']);
+        unset($agent['agent_rev']);
```
I would instead recommend to create a new model and get the version aware array from it. See https://github.com/fossology/fossology/pull/2617/files#diff-f62f1d2b860c504c87d386cdfa79513c109848593f0afeb60a5dc76e4b98ff42 for reference.

GMishx commented on src/www/ui/api/Middlewares/RestAuthMiddleware.php:
```diff
@@ -79,7 +79,8 @@ public function __invoke(Request $request, RequestHandler $handler) : ResponseIn
         throw new HttpForbiddenException("Do not have required scope.");
       }
-      $groupName = $request->getHeaderLine('groupName');
+      $query = $request->getQueryParams();
+      $groupName = $query["groupName"] ?? null;
```
Please keep it version aware.

GMishx commented on src/www/ui/api/Controllers/UploadController.php:
```diff
@@ -417,11 +417,11 @@ private function changeUpload($request, $response, $args, $isCopy)
+    $reqBody = $this->getParsedBody($request);
+    $uploadType = $reqBody['uploadType'] ?? null;
     if (empty($uploadType)) {
       $uploadType = 'vcs';
```
As per the OpenAPI doc, `uploadType` is a required field and thus, should throw appropriate exception if empty.

GMishx reviewed (changes requested):
Please support both version 1 and 2 of the API.

Akashsah2003 commented:
I have made the following changes @GMishx

- Added support for both Version1 and Version2 for apis.
- `uploadType` is now throwing bad request exception.
- Created models `Agent.php`, `SuccessfulAgent.php`, `EditedLicense.php` in order to support both API V1 and V2 responses.
- Modified Unit tests accordingly.

Akashsah2003 commented:
@GMishx I have made the commit so that tests pass. Kindly review the pr.

GMishx commented on src/www/ui/api/Controllers/UploadController.php:
```diff
@@ -400,7 +404,7 @@ private function changeUpload($request, $response, $args, $isCopy)
         !is_numeric($newFolderID = $request->getHeaderLine('folderId'))) {
       throw new HttpBadRequestException("folderId header should be an integer!");
-    }
+    };
```
Irrelevant changes, please revert.

GMishx commented on src/www/ui/api/Controllers/UploadController.php:
```diff
@@ -455,17 +478,10 @@ public function postUpload($request, $response, $args)
     if (array_key_exists("location", $reqBody)) {
       $locationObject = $reqBody["location"];
-    } elseif ($request->getHeaderLine('uploadType') != 'file') {
+    } elseif ($reqBody['uploadType'] != 'file') {
```
Better use `$uploadType` variable already assigned based on API version.

GMishx commented on src/www/ui/api/Models/Agent.php:
```diff
@@ -0,0 +1,194 @@
+  public function getArray($version=ApiVersion::V1)
+  {
+    if ($version == ApiVersion::V2) {
+        return [
+            'successfulAgents' => $this->getSuccessfulAgents(),
+            'uploadId' => $this->getUploadId(),
+            'agentName' => $this->getAgentName(),
+            'currentAgentId' => $this->getCurrentAgentId(),
+            'currentAgentRev' => $this->getCurrentAgentRev(),
+            'isAgentRunning' => $this->getIsAgentRunning()
+        ];
+    }
+    return [
+        'successfulAgents' => $this->getSuccessfulAgents(),
+        'uploadId' => $this->getUploadId(),
+        'agentName' => $this->getAgentName(),
+        'currentAgentId' => $this->getCurrentAgentId(),
+        'currentAgentRev' => $this->getCurrentAgentRev(),
+        'isAgentRunning' => $this->getIsAgentRunning()
+    ];
```
Please use 2 space indentation.

GMishx replied:
Both the responses (v1 and v2) looks same to me. Can the check be simply removed?

GMishx commented on src/www/ui/api/Models/EditedLicense.php:
```diff
@@ -0,0 +1,145 @@
+  public function getArray($version=ApiVersion::V1)
+  {
+    if ($version == ApiVersion::V2) {
+        return [
+            'id' => $this->getId(),
+            'shortName' => $this->getShortName(),
+            'count' => $this->getCount(),
+            'spdxId' => $this->getSpdxId()
+        ];
+    }
+    return [
+        'id' => $this->getId(),
+        'shortName' => $this->getShortName(),
+        'count' => $this->getCount(),
+        'spdx_id' => $this->getSpdxId()
+    ];
```
Please use 2 space indentation.

GMishx commented on src/www/ui/api/Models/SuccessfulAgent.php:
```diff
@@ -0,0 +1,122 @@
+  public function getArray($version=ApiVersion::V1)
+  {
+    if ($version == ApiVersion::V2) {
+        return [
+            'agentId' => $this->getAgentId(),
+            'agentRev' => $this->getAgentRev(),
+            'agentName' => $this->getAgentName()
+        ];
+    }
+    return [
+        'agent_id' => $this->getAgentId(),
+        'agent_rev' => $this->getAgentRev(),
+        'agent_name' => $this->getAgentName()
+    ];
```
Please use 2 space indentation.

GMishx reviewed (changes requested):
Minor changes, rest looks good.

Akashsah2003 commented:
Hi @GMishx I have made the required changes. Please take a look.

Akashsah2003 commented:
Hi @GMishx , Friendly reminder to review the latest changes in the PR. Your input is appreciated. Thanks!

GMishx reviewed (approved):
Changes looks good.

GMishx reviewed (approved):
Tested, working as expected.
